llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp

```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
   return PreservedAnalyses::none();
 }
 
+// Perform instrumentation sampling.
+// We transform:
+//   Increment_Instruction;
+// to:
+//   if (__llvm_profile_sampling__ <= SampleDuration) {
+//     Increment_Instruction;
+//   }
+//   __llvm_profile_sampling__ += 1;
```

Comment:
This seems more like `firstN of 64K` rather than random sampling. I have a couple of concerns.
1. This seems like it could easily introduce bias.
2. We introduce yet another knob for tuning `sampled-instr-duration`. If it was randomized then we won't have to worry about this knob. A fast random implementation may not be too much increase in overhead compared to this approach.

Reply:
Agree with what @snehasish said. Why not record once for every N hits? firstN of 64K is more prone to synchronization issues (with the program).

Reply:
xur@'s sampling scheme is bursty style which seems less likely to introduce bias compared with every N hits scheme (which can lead to shadow. 

For this style of sampling, there might be some optimization that can be done to coalesce sample count update and check.

Reply:
Since `__llvm_profile_sampling__` is shared by all PGO counters, I'm wondering if biasing could happen to code always shadowed in the range `[__llvm_profile_sampling__, 65536]`. E.g, assume two loops in the program, they count up to 65536 iterations in all, and the second loop will unlikely be counted?

Reply:
It can be problem for static count, but might be ok with runtime count. Data changes at runtime introduce some randomness -- for instance loop trip count. It is unlikely all loops in a program have fixed trip count through out the training run.

Reply:
__llvm_profile_sampling__ is not shared by all PGO counters. It's s thread-local variable. All the counters in one thread shared the value.

This is burst sampling. We used to have to parameters (i.e. changing 64K to another used specified value).  Using 64K value does increase the chances of bias. But I do think the chances are low in real programs.

Of course, I can write a test case to show the method will result in a bias profile. This is sampling after all. There will be always corner cases that gets biased result.

Reply:
> xur@'s sampling scheme is bursty style which seems less likely to introduce bias compared with every N hits scheme

Why? In general, sampling needs to be evenly distributed to avoid bias. HW sampling doesn't do burst either.  Do you have data and example to illustrate the benefit of bursty sampling? 

The only reason I can see to choose bursty sample is to open up the opportunity for coalescing increments, but that is not done here.